### PR TITLE
fstests: add TestFsPutChunked

### PR DESCRIPTION
**Pull request by B4dM4n**
#### What is the purpose of this change?

This adds a test for chunked uploads to the integration tests.
Each remote now has a `ChunkedUploadConfig` field in the `fstests.Opt` struct, to specify the required settings for the test. In addition, each remote that supports chunked uploads has to implement the `fstest.SetUploadChunkSizer` interface, which allows to change the value at runtime. Remotes that do not implement the `fstest.SetUploadChunkSizer` interface will not be tested, since the test would be the same as `TestFsPut` with bigger files.

These are the currently minimum chunk sizes and the total number of bytes uploaded to the remote during the test:

remote | min chunk size | test upload size
------ | -------------- | ----------------
b2 | 5M | 190M
s3 | 5M | 190M
drive | 256K | 13M
onedrive | 320K | 25M
\* | 1 | 28M

These numbers could probably be reduced a bit by removing some similar test cases. There should be no file download involved, as `file.Check` only compares the hash and modtime.

The output of the following tests can be viewed here: [chunked.log](https://github.com/ncw/rclone/files/2456792/chunked.log)

`go test -v ./backend/drive ./backend/onedrive ./backend/s3 ./backend/local -test.run '^TestIntegration/((Setup|Init|Purge|Finalise)|FsPutChunked)$' | tee chunked.log`

#### Was the change discussed in an issue or in the forum before?

#2549

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/ncw/rclone/blob/master/CONTRIBUTING.md#submitting-a-pull-request).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/ncw/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

### CODE REVIEW

**Comment by B4dM4n** [comment]
I had a few thoughts about the selection of chunk size values and came up with this list:
```go
		next := func(f func(fs.SizeSuffix) fs.SizeSuffix) fs.SizeSuffix {
			s := f(minChunkSize)
			if s > maxChunkSize {
				s = minChunkSize
			}
			return s
		}

		chunkSizes := fs.SizeSuffixList{
			minChunkSize,
			minChunkSize + (maxChunkSize-minChunkSize)/3,
			next(NextPowerOfTwo),
			next(NextMultipleOf(100000)),
			next(NextMultipleOf(100001)),
			maxChunkSize,
		}
		chunkSizes.Sort()
```

Besides the better range of edge values, this also reduces the uploaded data volume:

remote | min chunk size | test upload size
-- | -- | --
b2 | 5M | 139.10 M
s3 | 5M | 139.10 M
drive | 256K | 15.00 M
onedrive | 320K | 36.25 M
\* | 1 | 11.43 M

File: fstest/fstests/bits.go [hunk 1]
```diff
@@ -0,0 +1,33 @@
+//+build !go1.9
+
+package fstests
+
+func leadingZeros64(x uint64) int {
```
Comment (ncw) [inline fstest/fstests/bits.go]: How is this different from math.LeadingZeros64?

File: lib/readers/patter_reader_test.go [hunk 2]
```diff
@@ -0,0 +1,30 @@
+package readers
```
Comment (ncw) [inline lib/readers/patter_reader_test.go]: There is a typo in the filename here!  `patter_reader_test.go` should be `pattern_reader_test.go`

File: fstest/fstests/bits.go (hunk 1 above)
Comment (B4dM4n) [reply fstest/fstests/bits.go]: The [`bits.LeadingZeros64`](https://golang.org/pkg/math/bits/#LeadingZeros64) was added in go1.9, so a backport was needed for the tests.

The official version is probably faster, but performance is not key for the usage.

**Comment by B4dM4n** [comment]
The wrong filename is fixed and the `bits*.go` files are moved to the right commit.

Let's see how the integration tests turn out :smile:

File: fstest/fstests/bits.go (hunk 1 above)
Comment (ncw) [reply fstest/fstests/bits.go]: Ah, that makes sense!

**Comment by ncw** [comment]
Thanks :-)  Will merge - check https://pub.rclone.org/integration-tests/ tomorrow!

**Comment by ncw** [comment]
Looks like the tests were successful except for azureblob storage

```
   --- FAIL: TestIntegration/TestFsPutChunked (0.60s)
        --- FAIL: TestIntegration/TestFsPutChunked/100 (0.18s)
            --- PASS: TestIntegration/TestFsPutChunked/100/99 (0.05s)
            --- FAIL: TestIntegration/TestFsPutChunked/100/100 (0.06s)
                fstest.go:132: 
                    	Error Trace:	fstest.go:132
                    	            				fstest.go:138
                    	            				fstests.go:172
                    	            				fstests.go:572
                    	Error:      	Should be true
                    	Test:       	TestIntegration/TestFsPutChunked/100/100
                    	Messages:   	Azure container rclone-test-wilodeb4jafonec1cuvacuw9/chunked-100-100.bin: MD5 hash incorrect - expecting "7acedd1a84a4cfcb6e7a16003242945e" got "d41d8cd98f00b204e9800998ecf8427e"
                fstest.go:139: 
                    	Error Trace:	fstest.go:139
                    	            				fstests.go:172
                    	            				fstests.go:572
                    	Error:      	Not equal: 
                    	            	expected: 100
                    	            	actual  : 0
                    	Test:       	TestIntegration/TestFsPutChunked/100/100
                    	Messages:   	chunked-100-100.bin: size incorrect file=100 vs obj=0
                fstest.go:132: 
                    	Error Trace:	fstest.go:132
                    	            				fstest.go:138
                    	            				fstests.go:175
                    	            				fstests.go:572
                    	Error:      	Should be true
                    	Test:       	TestIntegration/TestFsPutChunked/100/100
                    	Messages:   	Azure container rclone-test-wilodeb4jafonec1cuvacuw9/chunked-100-100.bin: MD5 hash incorrect - expecting "7acedd1a84a4cfcb6e7a16003242945e" got "d41d8cd98f00b204e9800998ecf8427e"
                fstest.go:139: 
                    	Error Trace:	fstest.go:139
                    	            				fstests.go:175
                    	            				fstests.go:572
                    	Error:      	Not equal: 
                    	            	expected: 100
                    	            	actual  : 0
                    	Test:       	TestIntegration/TestFsPutChunked/100/100
                    	Messages:   	chunked-100-100.bin: size incorrect file=100 vs obj=0
            --- PASS: TestIntegration/TestFsPutChunked/100/201 (0.07s)
        --- FAIL: TestIntegration/TestFsPutChunked/128 (0.15s)
            --- PASS: TestIntegration/TestFsPutChunked/128/127 (0.04s)
            --- FAIL: TestIntegration/TestFsPutChunked/128/128 (0.05s)
                fstest.go:132: 
                    	Error Trace:	fstest.go:132
                    	            				fstest.go:138
                    	            				fstests.go:172
                    	            				fstests.go:572
                    	Error:      	Should be true
                    	Test:       	TestIntegration/TestFsPutChunked/128/128
                    	Messages:   	Azure container rclone-test-wilodeb4jafonec1cuvacuw9/chunked-128-128.bin: MD5 hash incorrect - expecting "37eff01866ba3f538421b30b7cbefcac" got "d41d8cd98f00b204e9800998ecf8427e"
                fstest.go:139: 
                    	Error Trace:	fstest.go:139
                    	            				fstests.go:172
                    	            				fstests.go:572
                    	Error:      	Not equal: 
                    	            	expected: 128
                    	            	actual  : 0
                    	Test:       	TestIntegration/TestFsPutChunked/128/128
                    	Messages:   	chunked-128-128.bin: size incorrect file=128 vs obj=0
                fstest.go:132: 
                    	Error Trace:	fstest.go:132
                    	            				fstest.go:138
                    	            				fstests.go:175
                    	            				fstests.go:572
                    	Error:      	Should be true
                    	Test:       	TestIntegration/TestFsPutChunked/128/128
                    	Messages:   	Azure container rclone-test-wilodeb4jafonec1cuvacuw9/chunked-128-128.bin: MD5 hash incorrect - expecting "37eff01866ba3f538421b30b7cbefcac" got "d41d8cd98f00b204e9800998ecf8427e"
                fstest.go:139: 
                    	Error Trace:	fstest.go:139
                    	            				fstests.go:175
                    	            				fstests.go:572
                    	Error:      	Not equal: 
                    	            	expected: 128
                    	            	actual  : 0
                    	Test:       	TestIntegration/TestFsPutChunked/128/128
                    	Messages:   	chunked-128-128.bin: size incorrect file=128 vs obj=0
            --- PASS: TestIntegration/TestFsPutChunked/128/257 (0.06s)
        --- FAIL: TestIntegration/TestFsPutChunked/133 (0.14s)
            --- PASS: TestIntegration/TestFsPutChunked/133/132 (0.04s)
            --- FAIL: TestIntegration/TestFsPutChunked/133/133 (0.05s)
                fstest.go:132: 
                    	Error Trace:	fstest.go:132
                    	            				fstest.go:138
                    	            				fstests.go:172
                    	            				fstests.go:572
                    	Error:      	Should be true
                    	Test:       	TestIntegration/TestFsPutChunked/133/133
                    	Messages:   	Azure container rclone-test-wilodeb4jafonec1cuvacuw9/chunked-133-133.bin: MD5 hash incorrect - expecting "8b6831066bd6fa5d47714f2ea8bd137e" got "d41d8cd98f00b204e9800998ecf8427e"
                fstest.go:139: 
                    	Error Trace:	fstest.go:139
                    	            				fstests.go:172
                    	            				fstests.go:572
                    	Error:      	Not equal: 
                    	            	expected: 133
                    	            	actual  : 0
                    	Test:       	TestIntegration/TestFsPutChunked/133/133
                    	Messages:   	chunked-133-133.bin: size incorrect file=133 vs obj=0
                fstest.go:132: 
                    	Error Trace:	fstest.go:132
                    	            				fstest.go:138
                    	            				fstests.go:175
                    	            				fstests.go:572
                    	Error:      	Should be true
                    	Test:       	TestIntegration/TestFsPutChunked/133/133
                    	Messages:   	Azure container rclone-test-wilodeb4jafonec1cuvacuw9/chunked-133-133.bin: MD5 hash incorrect - expecting "8b6831066bd6fa5d47714f2ea8bd137e" got "d41d8cd98f00b204e9800998ecf8427e"
                fstest.go:139: 
                    	Error Trace:	fstest.go:139
                    	            				fstests.go:175
                    	            				fstests.go:572
                    	Error:      	Not equal: 
                    	            	expected: 133
                    	            	actual  : 0
                    	Test:       	TestIntegration/TestFsPutChunked/133/133
                    	Messages:   	chunked-133-133.bin: size incorrect file=133 vs obj=0
            --- PASS: TestIntegration/TestFsPutChunked/133/267 (0.05s)
        --- FAIL: TestIntegration/TestFsPutChunked/200 (0.14s)
            --- PASS: TestIntegration/TestFsPutChunked/200/199 (0.04s)
            --- FAIL: TestIntegration/TestFsPutChunked/200/200 (0.05s)
                fstest.go:132: 
                    	Error Trace:	fstest.go:132
                    	            				fstest.go:138
                    	            				fstests.go:172
                    	            				fstests.go:572
                    	Error:      	Should be true
                    	Test:       	TestIntegration/TestFsPutChunked/200/200
                    	Messages:   	Azure container rclone-test-wilodeb4jafonec1cuvacuw9/chunked-200-200.bin: MD5 hash incorrect - expecting "fb7001d34b8e82c9b579be5005d5b0a5" got "d41d8cd98f00b204e9800998ecf8427e"
                fstest.go:139: 
                    	Error Trace:	fstest.go:139
                    	            				fstests.go:172
                    	            				fstests.go:572
                    	Error:      	Not equal: 
                    	            	expected: 200
                    	            	actual  : 0
                    	Test:       	TestIntegration/TestFsPutChunked/200/200
                    	Messages:   	chunked-200-200.bin: size incorrect file=200 vs obj=0
                fstest.go:132: 
                    	Error Trace:	fstest.go:132
                    	            				fstest.go:138
                    	            				fstests.go:175
                    	            				fstests.go:572
                    	Error:      	Should be true
                    	Test:       	TestIntegration/TestFsPutChunked/200/200
                    	Messages:   	Azure container rclone-test-wilodeb4jafonec1cuvacuw9/chunked-200-200.bin: MD5 hash incorrect - expecting "fb7001d34b8e82c9b579be5005d5b0a5" got "d41d8cd98f00b204e9800998ecf8427e"
                fstest.go:139: 
                    	Error Trace:	fstest.go:139
                    	            				fstests.go:175
                    	            				fstests.go:572
                    	Error:      	Not equal: 
                    	            	expected: 200
                    	            	actual  : 0
                    	Test:       	TestIntegration/TestFsPutChunked/200/200
                    	Messages:   	chunked-200-200.bin: size incorrect file=200 vs obj=0
            --- PASS: TestIntegration/TestFsPutChunked/200/401 (0.05s)
```

@B4dM4n & @sandeepkru any ideas?

**Comment by B4dM4n** [comment]
I can't tell what is wrong there. All the tests that fail are `chunkSize = fileSize`. A quick glance at the azureblob code has not revealed any problems.

**Comment by ncw** [comment]
OK I'll do some manual tests - this might be a release blocker :-(

**Comment by ncw** [comment]
OK it turns out that this is a bug in the azureblob SDK:

https://github.com/Azure/azure-storage-blob-go/pull/75

Also the new test framework isn't setting upload-cutoff which means the uploads aren't multipart on some backends!  I'll fix that in a moment!

**Comment by B4dM4n** [comment]
I pushed a small follow-up to this PR, as I noticed the azureblob test only used a `200 B` upper chunk size instead of the intended `2 MB`: 15ce0ae57cf52990e10fdb62a6303a9023db7e91

**Comment by ncw** [comment]
Thanks @B4dM4n - I noticed that too (but I thought it was deliberate!)
